### DISCUSSION TITLE
UI needs

### DISCUSSION THREAD

**Discussion by agracey** [epinio/epinio#1150] [Ideas]
Here's a discussion for the [UI milestone](https://github.com/epinio/epinio/milestone/13)

**Comment by agracey** [comment]
Epinio UI stories


Separate audiences with different goals but similar needs:
- Devs
- Ops
- Maybe compliance?

Dev Goals:
- Discoverability of platform capabilities
- Deployment of app without CLI 
- Services set up
- - Marketplace
- Tailing logs / observability
- Explore under the hood 
- Track namespace resource allocation (and usage against that allocation)


Dev nice to haves:

- code-server
- view traffic between applications (and services)
- Shell access via UI
- Know how the build happened and which buildpacks in the builder got run
- - Control which buildpacks get run
- Open API spec generation based on network DPI


Ops Goals:
- What are my devs running? And who is running it?
- Platform configuration (as possible)
- - s3/registry
- - User access (eventually RBAC)
- - Default helm chart being deployed for apps (assuming we make this change)
- - Service Marketplace set up 
- Provide custom services to appropiate namespaces
- Deploy applications from git
- - And allow for updates 
- - Basically, treat as "black box CD"
- Identifying what services are used (and which versions so they can schedule maintenance appropriately)
- View Epinio clusters and apps running across all environments

Ops nice to haves:
- Automatic update of app as git changes
- Promotion between clusters
- Security audit of running apps

**Reply by kkaempf** [reply]
Can you prioritize these ?

**Reply by kkaempf** [reply]
What's the single most important feature you can't live without ?

**Reply by kkaempf** [reply]
In which role (Dev, Ops, Compliance) is the majority of our user base ?
Who's sales talking to ?
Who's taking the buy/no-buy decision ?

**Comment by richard-cox** [comment]
Some of these will be out of scope, but for high level requirements I'd like to add. Some also cross over with sections from above

- Ops should be able to manage resource usage of and access to namespaces per user or group (for example a github org)
  - Namespace RBAC
  - Namespace quotas 
- Ops should be able to review the state of an epinio instance to understand it's usage (resources consumed, app states, events, etc)
  - This includes more granular views for specific namespaces
- Users should be able to quickly review the state of their applications and give attention where needed
  - Are any down? Has one been going constantly down? This could be achieved by an application health check 
  - Are any resource starved?
- Users should be able to create an application
  - Easily guide the user from source to a running application
  - Source can come from supplied files (local file/s), remote files (public/private github/gitlab) or pre-built files (container image)
  - Custom routes and new services can be created during the application create flow
- Users should be able to review in detail the state of their application
  - What's the current output (logs) and events?
  - Is it 'healthy', what resources is it consuming out of it's max allowed? 
  - When was it created, deployed?
  - Where did the source come from?
  - How was it built? 
  - What resources are associated with it (show the bound services, routes, application 'instances', etc)?
  - What is the state of the `instances`, what resources are they consuming? How log have they been up for? What are the logs?
  - How can the user access the app via the CLI?
- Users should be able to manage an existing application
  - Stop/Start the application
  - Restart a deployed app
  - Stage and redeploy from a different source  
  - Terminate an application instance
  - SSH to an application instance

The big one's from my perspective which we should tackle first are

- Logs
  - Application 
  - Application Instance
- SSH
  - Application Instance
- Events
  - Cluster
  - Namespace
  - Application  
 - Github/Gitlab Integration
   - private github/gitlab would require a way to register instances and credentials
   - public github/gitlab might also benefit from pre-registered credentials (to avoid API rate limiting)
   - The UI would use instance/creds to fetch repo/branch/commit info
   - The epinio backend would need to fetch the code using repo & commit info before continuing with the normal process
   - The user should be able to select a commit of an existing app and redeploy from it

Additionally
- Application Health Check
- Namespace and application quotas

**Comment by agracey** [comment]
Oh, users should also be able to download the config file to allow access to Epinio via the UI
